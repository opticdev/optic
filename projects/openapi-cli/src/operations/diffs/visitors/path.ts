import { OpenAPIV3 } from '../../../specs';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { PathComponents, PathComponent } from '../..';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };
  }

  // TODO: match path's parameters against those documented
  // let parameterComponents = PathComponents.fromPath(path).filter(
  //   PathComponent.isTemplate
  // );
  // const spec = specOption.unwrap()
  // spec.parameters
}
