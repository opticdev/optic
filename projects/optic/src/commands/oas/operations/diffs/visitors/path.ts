import { OpenAPIV3 } from '../../../specs';
import {
  OperationDiffResult,
  OperationDiffResultKind,
} from '../../../../capture/patches/patchers/spec/types';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };
  }

  // TODO: match path's parameters against those documented
  // let parameterComponents = PathComponents.fromPath(path).filter(
  //   PathComponent.isTemplate
  // );
  // const spec = specOption.unwrap()
  // spec.parameters
}
