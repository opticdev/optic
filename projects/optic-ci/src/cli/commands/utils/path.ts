import path from 'path';
import { promisify } from 'util';
import { exec as callbackExec } from 'child_process';
const exec = promisify(callbackExec);

export const getRelativeRepoPath = (
  relativePath: string,
  gitRootPath: string | false
): string => {
  const absolutePath = path.join(process.cwd(), relativePath);

  if (!gitRootPath) {
    return absolutePath;
  } else {
    return absolutePath.startsWith(gitRootPath)
      ? absolutePath.replace(gitRootPath, '')
      : absolutePath;
  }
};

export const getGitRootPath = async () => {
  const { stdout: gitRootUntrimmed } = await exec(
    'git rev-parse --show-toplevel'
  );
  const gitRoot = gitRootUntrimmed.trim();
  return gitRoot;
};
