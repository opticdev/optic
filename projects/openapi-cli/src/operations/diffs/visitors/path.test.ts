import { visitPath } from './path';
import { Some, None } from 'ts-results';

describe('visitPath', () => {
  it('detects an unspecified path', () => {
    const matchingResults = [
      ...visitPath('/orders/{orderId}', Some({}), {
        pathPattern: Some('/orders/{orderId}'),
      }),
    ];
    expect(matchingResults).toHaveLength(0);

    const unmatchingResults = [
      ...visitPath('/orders/{orderId}', None, { pathPattern: None }),
    ];
    expect(unmatchingResults).toHaveLength(1);
    expect(unmatchingResults).toMatchSnapshot();
  });
});
