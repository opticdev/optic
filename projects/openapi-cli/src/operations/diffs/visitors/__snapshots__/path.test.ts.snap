// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visitPath detects an unspecified path 1`] = `
Array [
  Object {
    "kind": "UnmatchedPath",
    "subject": "/orders/{orderId}",
  },
]
`;
